spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-fabric8-client-catalog-watcher/src/test/java/org/springframework/cloud/kubernetes/fabric8/catalog/watch/CatalogWatchIT.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2013-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.kubernetes.fabric8.catalog.watch;
+
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.fabric8.kubernetes.api.model.networking.v1.Ingress;
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.k3s.K3sContainer;
+import reactor.netty.http.client.HttpClient;
+import reactor.util.retry.Retry;
+import reactor.util.retry.RetryBackoffSpec;
+
+import org.springframework.cloud.kubernetes.commons.discovery.EndpointNameAndNamespace;
+import org.springframework.cloud.kubernetes.integration.tests.commons.Commons;
+import org.springframework.cloud.kubernetes.integration.tests.commons.Fabric8Utils;
+import org.springframework.cloud.kubernetes.integration.tests.commons.K8SUtils;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.core.ResolvableType;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.client.reactive.ReactorClientHttpConnector;
+import org.springframework.web.reactive.function.client.WebClient;
+
+import static org.awaitility.Awaitility.await;
+
+/**
+ * @author wind57
+ */
+class CatalogWatchIT {
+
+	private static final String APP_NAME = "spring-cloud-kubernetes-fabric8-client-catalog-watcher";
+
+	private static final String NAMESPACE = "default";
+
+	private static final K3sContainer K3S = Commons.container();
+
+	private static KubernetesClient client;
+
+	private String busyboxServiceName;
+
+	private String busyboxDeploymentName;
+
+	private String appDeploymentName;
+
+	private String appServiceName;
+
+	private String appIngressName;
+
+	@BeforeAll
+	static void beforeAll() {
+		K3S.start();
+		Config config = Config.fromKubeconfig(K3S.getKubeConfigYaml());
+		client = new DefaultKubernetesClient(config);
+	}
+
+	/**
+	 * <pre>
```

Comment:
this is the idea of the test - nothing fancy, really. But I felt that we need a proper integration test here...